File: apps/web/src/api/holdings.ts
import { z } from "zod";

/** Relative, because the dev server proxies /api to the API on the same origin. */
export const HOLDINGS_URL = "/api/holdings";

/** Unrealised results — and a broken ledger's quantity — may be negative. */
export const SIGNED_DECIMAL_PATTERN = /^-?\d+(\.\d+)?$/;

const signedDecimal = z.string().regex(SIGNED_DECIMAL_PATTERN);

/**
 * One (Account, Instrument) holding (ticket 20). Every figure is EUR — the
 * DisplayCurrency conversion happens client-side, for presentation only.
 * `marker` names why a position is excluded from totals; `basis_gap` names
 * why the basis cannot be stated while the position itself still counts.
 */
export const positionSchema = z.object({
  instrument_id: z.number(),
  symbol: z.string(),
  name: z.string(),
  family: z.enum(["crypto", "security", "cash"]),
  type: z.string(),
  chain: z.string().nullable(),
  contract_address: z.string().nullable(),
  isin: z.string().nullable(),
  is_numeraire: z.boolean(),
  account_id: z.number(),
  account_name: z.string(),
  access_software: z.string().nullable(),
  platform_name: z.string(),
  platform_kind: z.enum(["exchange", "cold_storage", "software_wallet", "broker", "bank"]),
  quantity: signedDecimal,
  marker: z.enum(["dangerous", "ignored", "unacknowledged", "unpriced"]).nullable(),
  basis_eur: signedDecimal.nullable(),
  basis_gap: z.enum(["awaiting_valuation", "unvouched"]).nullable(),
  average_cost_eur: signedDecimal.nullable(),
  value_eur: signedDecimal.nullable(),
  unrealised_eur: signedDecimal.nullable(),
  price_source: z.string().nullable(),
  price_as_of: z.iso.datetime({ offset: true }).nullable(),
  rate_date: z.iso.date().nullable(),
});

export const holdingsSchema = z.object({ positions: z.array(positionSchema) });

/**
 * The latest published reference rate for a DisplayCurrency — units per one
 * euro, with the date it represents. Presentation only, never a tax figure.
 */
export const displayRateSchema = z.object({
  currency: z.string(),
  rate: signedDecimal,
  rate_date: z.iso.date(),
});

export type Position = z.infer<typeof positionSchema>;
export type Holdings = z.infer<typeof holdingsSchema>;
export type DisplayRate = z.infer<typeof displayRateSchema>;

export async function fetchHoldings(): Promise<Holdings> {
  const response = await fetch(HOLDINGS_URL);
  if (!response.ok) {
    throw new Error(`The API answered ${response.status} instead of the holdings.`);
  }
  return holdingsSchema.parse(await response.json());
}

export async function fetchDisplayRate(currency: string): Promise<DisplayRate> {
  const response = await fetch(`${HOLDINGS_URL}/display-rate/${currency}`);
  if (!response.ok) {
    throw new Error(`The API answered ${response.status} instead of the ${currency} rate.`);
  }
  return displayRateSchema.parse(await response.json());
}

File: apps/api/src/open_leprechaun/routers/holdings.py
from datetime import date
from decimal import Decimal
from typing import Annotated, Literal

from fastapi import APIRouter, HTTPException
from pydantic import AwareDatetime, BaseModel, PlainSerializer

from open_leprechaun.auth import AdminDep
from open_leprechaun.db import EngineDep
from open_leprechaun.rates import ReferenceRateSourceDep
from open_leprechaun.services import holdings
from open_leprechaun.services.holdings import Position, RateUnavailableError

router = APIRouter(tags=["holdings"])

# Fixed-point on the way out, as everywhere: a quantity or an EUR figure is
# answered as a decimal string, never a float.
Fixed = Annotated[Decimal, PlainSerializer(lambda value: format(value, "f"), return_type=str)]


class PositionResponse(BaseModel):
    instrument_id: int
    symbol: str
    name: str
    family: str
    type: str
    chain: str | None
    contract_address: str | None
    isin: str | None
    is_numeraire: bool
    account_id: int
    account_name: str
    access_software: str | None
    platform_name: str
    platform_kind: str
    quantity: Fixed
    # Why a position is excluded from totals; None means it counts.
    marker: Literal["dangerous", "ignored", "unacknowledged", "unpriced"] | None
    basis_eur: Fixed | None
    # Why the basis cannot be stated while the position still counts.
    basis_gap: Literal["awaiting_valuation", "unvouched"] | None
    average_cost_eur: Fixed | None
    value_eur: Fixed | None
    unrealised_eur: Fixed | None
    price_source: str | None
    price_as_of: AwareDatetime | None
    rate_date: date | None

    @classmethod
    def of(cls, position: Position) -> PositionResponse:
        return cls(**vars(position))


class HoldingsResponse(BaseModel):
    positions: list[PositionResponse]


class DisplayRateResponse(BaseModel):
    currency: str
    # Units of currency per one euro, as published.
    rate: Fixed
    rate_date: date


@router.get(
    "/holdings",
    summary="Everything held — crypto and cash — with cost basis from the Tax Lots"
    " and values from the store, in EUR",
    response_model=HoldingsResponse,
)
def list_holdings(admin: AdminDep, engine: EngineDep) -> HoldingsResponse:
    return HoldingsResponse(
        positions=[PositionResponse.of(position) for position in holdings.portfolio(engine)]
    )


@router.get(
    "/holdings/display-rate/{currency}",
    summary="The latest reference rate for a DisplayCurrency — presentation only,"
    " never a tax figure",
    response_model=DisplayRateResponse,
)
def get_display_rate(
    currency: str, admin: AdminDep, engine: EngineDep, source: ReferenceRateSourceDep
) -> DisplayRateResponse:
    try:
        rate = holdings.display_rate(engine, source, currency=currency.upper())
    except RateUnavailableError as unavailable:
        raise HTTPException(status_code=404, detail=str(unavailable)) from unavailable
    return DisplayRateResponse(currency=currency.upper(), rate=rate.rate, rate_date=rate.rate_date)
